### DISCUSSION TITLE
RUN button: to know when the execution is finished

### DISCUSSION THREAD

**Discussion by ThomRoy** [pyscript/pyscript#1418] [Q&A]
Hello everybody!

I added «.py-repl-run-button{opacity:1}» to make the RUN button visible.

But is it possible to make it change its appearance so that we know when the execution is finished (like in Google Colab for example)?

Thanks for your help!

**Comment by JeffersGlass** [comment]
This is currently possible using a feature which has been merged but is not a part of any release yet: plugin hooks for py-repl. (#1106) These allow you to execute before and after any REPL execution.

You can try out this feature by linking to the `unstable` branch of PyScript, but **be warned** - that branch is called unstable for a reason, things will change and break on it before the next release. I mention it if you want to *play around* with this feature, not treat it as stable.

Here's an example using a plugin to change the opacity and color of the button of the currently executing REPL, while it is running:

```html
<!-- index.html -->
<script defer src="https://pyscript.net/unstable/pyscript.js"></script>
<link rel="stylesheet" href="https://pyscript.net/unstable/pyscript.css">

<py-config>
    plugins = ['replstatus.py']
</py-config>
<py-repl>
    #Some long-running sample code, so you can see the run button change colors
    print("start")
    for x in range(10_000_000):
        y = 1

    print("done")
</py-repl>
```
```python
#replstatus.py
import asyncio

class ReplStatus():
    def init(self, *args):
        ...

    async def beforePyReplExec(self, *, interpreter, src, outEl, pyReplTag):
        await self.setRunButtonState(pyReplTag, running=True)

    async def afterPyReplExec(self, *, interpreter, src, outEl, pyReplTag, result):
        await self.setRunButtonState(pyReplTag, running=False)

    async def setRunButtonState(self, pyReplTag, running):
        button = pyReplTag.querySelector('.py-repl-run-button')
        svg = button.querySelector('svg')

        if running:
            button.style.opacity = 1
            svg.style.color = 'red'
        else:
            button.style.opacity = ''
            svg.style.color = 'green'

        await asyncio.sleep(0.01) # Yield to browser event loop


plugin = ReplStatus()
```

**Comment by ThomRoy** [comment]
Thanks a lot for your answer!

I tried to replace by loading the unstable version of Pyscript (and adding replstatus.py and loading it as a plugin), but the splashscreen is still stuck at :
```
Downloading pyodide-0.22.1...
Python startup...
Python ready!
Setting up virtual environment...
```
Indeed, it doesn't seem to be usable at all for the moment! 😅

**Comment by CarlKCarlK** [comment]
@JeffersGlass, as you warned this has stopped working. Is there an update (stable or not). Thanks! - Carl

My physics demo: https://carlkcarlk.github.io/perfect-physics/pyscript/

**Reply by JeffersGlass** [reply]
@madhur-tandon can you shed some light on this? It appears the pyReplTag that's being passed to `beforePyReplExec`/`afterPyReplExec` is still not being properly proxied since #1490 was merged. This is the [exact issue we were discussing here](https://github.com/pyscript/pyscript/pull/1490/files#diff-f59eb1d30675fcb8f60f95180de11ef04f3bdfe207a1432eae89c392a04d3c78) (attribute access on those HTML elements being broken).

I had thought #1529 had fixed all of these issues, but perhaps not. In particular, restoring the changes to `test_plugins` to what that file was before 1490, both `test_pyscript_exec_hooks` and `test_pyrepl_exec_hooks` (which rely on that proxying working correctly) fail.
